docs/userguide/mpi_apps.rst

```diff
@@ -40,116 +73,275 @@ to the desired number of blocks.
-                    cpus_per_node=64,
+            executors=[
+                HighThroughputExecutor(
+                    label="htex",
+                    enable_mpi_mode=True,
+                    mpi_launcher="mpiexec",
+                    cores_per_worker=user_opts["cores_per_worker"],
+                    address=address_by_interface("bond0"),
+                    provider=PBSProProvider(
+                        launcher=SimpleLauncher(),
+                        account=user_opts["account"],
+                        queue=user_opts["queue"],
+                        # PBS directives (header lines): for array jobs pass '-J' option
+                        scheduler_options=user_opts["scheduler_options"],
+                        # Command to be run before starting a worker, such as:
+                        worker_init=user_opts["worker_init"],
+                        # number of compute nodes allocated for each block
+                        nodes_per_block=user_opts["nodes_per_block"],
+                        init_blocks=1,
+                        min_blocks=0,
+                        max_blocks=1, # Can increase more to have more parallel jobs
+                        walltime=user_opts["walltime"]
+                    ),
                 ),
-            ),
-        ]
-    )
+            ],
 
 
 Writing MPI-Compatible Apps
 ++++++++++++++++++++++++++++
 
-The `App <apps.html>`_ can be either a Python or Bash App which invokes the MPI application.
-
-In the easiest case (i.e., single MPI task per block), write the MPI launcher options in the string returned by
-the bash app or as part of a subprocess call from a Python app.
+In MPI mode, the :class:`~parsl.executors.high_throughput.executor.HighThroughputExecutor` can execute both Python or Bash Apps which invokes the MPI application.
+However, it is important to not that Python Apps that directly use ``mpi4py`` is not supported.
 
+For multi-node MPI applications, especially when running multiple applications within a single batch job,
+it is important to specify the resource requirements for the app so that the Parsl worker can provision
+the appropriate resources before the application starts. For eg, your Parsl script might contain a molecular
+dynamics application that requires 8 ranks over 1 node for certain inputs and 32 ranks over 4 nodes for some
+depending on the size of the molecules being simulated. By specifying resources via ``resource_specification``,
+parsl workers will provision the requested resources and then compose MPI launch command prefixes
+(Eg: ``mpiexec -n <ranks> -ppn <ranks_per_node> -hosts <node1..nodeN>``). These launch command prefixes are
+shared with the app via environment variables.
 
 .. code-block:: python
 
     @bash_app
-    def echo_hello(n: int, stderr='std.err', stdout='std.out'):
-        return f'mpiexec -n {n} --ppn 1 hostname'
+    def echo_hello(n: int, stderr='std.err', stdout='std.out', parsl_resource_specification: Dict):
```

Comment:
Do my functions need a "parsl_resource_specificiation" kwarg to be used in MPI mode?

Reply:
I'd strongly lean towards an implementation that does not require injecting Parsl logic into the underlying functions whenever possible.

Reply:
@yadudoc / @benclifford , what do you see as the remedy to this? It seems we go wrong from the beginning, in that the resources are combined with the "app kwargs" starting from the initial call to PythonApp.__call__ but then pulled out and treated separately throughout the Executor logic (e.g., as a separate argument in HTEX.submit). 

Could we not make it so that HTEx only adds the resource spec into the function call if it is present in the signature?

Reply:
There are no python apps involved in this MPI use case.

I think separately there's a reasonable use case that python functions can come from external sources and be decorated directly with `@python_app`; and Parsl could make it so those don't get passed parsl-specific kwargs.

That's out of scope for this MPI PR and I encourage you to write up an issue with the desired semantics separately.

Reply:
Alright, I'm convinced that we should punt on this issue so that we have the MPI logic usable with bash apps. I still think this issue should be resolved, and I'll make an Issue about it

Reply:
Marking this as resolved with the idea that we'll deal with the `parsl_resource_specification` handling in a separate issue.